### ISSUE TITLE
Locking does not work

### ISSUE THREAD

**Issue by davidak** [elementary/gala#1214]
## What Happened
When i click "Lock" in the power menu nothing happens.

## Expected Behavior
The session get's locked. Like in elementary OS 5.

## Steps to Reproduce

1. click "Lock" in the power menu
2. nothing happens

### Logs
nothing in the logs. not even with debugging output enabled

## Platform Information
elementaryos-6.0-stable.20210810.iso live mode
also on nixos which uses the latest releases of every package (from https://releases.elementary.io/)
maybe on nixos it is an issue that i updated it from release 5 packages? should i reset some settings?

does locking work when elementary OS 6 is installed?

it would be a security issue when you leave the computer unlocked

any way i can debug it on nixos? can i trigger the locking from console?

**Comment by davidmhewitt** [comment]
It seems to work on an installed system.

The lock call goes via the `org.gnome.ScreenSaver` D-Bus implementation, which is implemented in Gala. You could monitor the D-Bus traffic, or try manually firing that method.

**Comment by davidak** [comment]
is it maybe intended in the live session?

>You could monitor the D-Bus traffic

i tried that with `sudo dbus-monitor --system` but see nothing when i click "lock"

when i click the power menu i get these lines

<details>
  <summary>show</summary>
    
  ```
method call time=1629188212.847592 sender=:1.65 -> destination=org.freedesktop.login1 serial=205 path=/org/freedesktop/login1; interface=org.freedesktop.login1.Manager; member=ListUsers
method return time=1629188212.847736 sender=:1.4 -> destination=:1.65 serial=255 reply_serial=205
   array [
      struct {
         uint32 1000
         string "davidak"
         object path "/org/freedesktop/login1/user/_1000"
      }
   ]
method call time=1629188212.848103 sender=:1.65 -> destination=org.freedesktop.DBus serial=206 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=AddMatch
   string "type='signal',sender='org.freedesktop.login1',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged',path='/org/freedesktop/login1/user/_1000',arg0='org.freedesktop.login1.User'"
method return time=1629188212.848115 sender=org.freedesktop.DBus -> destination=:1.65 serial=150 reply_serial=206
method call time=1629188212.848119 sender=:1.65 -> destination=org.freedesktop.DBus serial=207 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=AddMatch
   string "type='signal',sender='org.freedesktop.login1',interface='org.freedesktop.login1.User',path='/org/freedesktop/login1/user/_1000'"
method return time=1629188212.848126 sender=org.freedesktop.DBus -> destination=:1.65 serial=151 reply_serial=207
method call time=1629188212.848163 sender=:1.65 -> destination=org.freedesktop.DBus serial=208 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=StartServiceByName
   string "org.freedesktop.login1"
   uint32 0
method return time=1629188212.848171 sender=org.freedesktop.DBus -> destination=:1.65 serial=152 reply_serial=208
   uint32 2
method call time=1629188212.848719 sender=:1.65 -> destination=org.freedesktop.DBus serial=209 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=GetNameOwner
   string "org.freedesktop.login1"
method return time=1629188212.848729 sender=org.freedesktop.DBus -> destination=:1.65 serial=153 reply_serial=209
   string ":1.4"
method call time=1629188212.848938 sender=:1.65 -> destination=:1.4 serial=210 path=/org/freedesktop/login1/user/_1000; interface=org.freedesktop.DBus.Properties; member=GetAll
   string "org.freedesktop.login1.User"
method return time=1629188212.849070 sender=:1.4 -> destination=:1.65 serial=256 reply_serial=210
   array [
      dict entry(
         string "UID"
         variant             uint32 1000
      )
      dict entry(
         string "GID"
         variant             uint32 100
      )
      dict entry(
         string "Name"
         variant             string "davidak"
      )
      dict entry(
         string "Timestamp"
         variant             uint64 1629185720727732
      )
      dict entry(
         string "TimestampMonotonic"
         variant             uint64 10835793
      )
      dict entry(
         string "RuntimePath"
         variant             string "/run/user/1000"
      )
      dict entry(
         string "Service"
         variant             string "user@1000.service"
      )
      dict entry(
         string "Slice"
         variant             string "user-1000.slice"
      )
      dict entry(
         string "Display"
         variant             struct {
               string "2"
               object path "/org/freedesktop/login1/session/_32"
            }
      )
      dict entry(
         string "State"
         variant             string "active"
      )
      dict entry(
         string "Sessions"
         variant             array [
               struct {
                  string "2"
                  object path "/org/freedesktop/login1/session/_32"
               }
            ]
      )
      dict entry(
         string "IdleHint"
         variant             boolean false
      )
      dict entry(
         string "IdleSinceHint"
         variant             uint64 0
      )
      dict entry(
         string "IdleSinceHintMonotonic"
         variant             uint64 0
      )
      dict entry(
         string "Linger"
         variant             boolean false
      )
   ]
method call time=1629188212.849744 sender=:1.65 -> destination=org.freedesktop.DBus serial=211 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=RemoveMatch
   string "type='signal',sender='org.freedesktop.login1',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged',path='/org/freedesktop/login1/user/_1000',arg0='org.freedesktop.login1.User'"
method return time=1629188212.849757 sender=org.freedesktop.DBus -> destination=:1.65 serial=154 reply_serial=211
method call time=1629188212.849760 sender=:1.65 -> destination=org.freedesktop.DBus serial=212 path=/org/freedesktop/DBus; interface=org.freedesktop.DBus; member=RemoveMatch
   string "type='signal',sender='org.freedesktop.login1',interface='org.freedesktop.login1.User',path='/org/freedesktop/login1/user/_1000'"
method return time=1629188212.849766 sender=org.freedesktop.DBus -> destination=:1.65 serial=155 reply_serial=212
  ```
  
</details>

>or try manually firing that method

how would i do that?

**Comment by davidmhewitt** [comment]
It's a session bus call, so you'd need to monitor the session bus.

To manually fire D-Bus calls, I like to use D-Feet to do it in a graphical way, but there are CLI options. I just don't know the syntax off the top of my head.

**Comment by davidak** [comment]
>It's a session bus call, so you'd need to monitor the session bus.

ok. i tried this, but no output when clicking

`sudo dbus-monitor --session`

>I like to use D-Feet

looks nice. can you tell me how to trigger it there?

![Bildschirmfoto von 2021-08-17 10-34-52](https://user-images.githubusercontent.com/91113/129693171-c0eedada-5594-495d-89f4-e49da4498327.png)

**Comment by davidmhewitt** [comment]
Expand out the `/org/gnome/ScreenSaver` section, there should be a `Lock ()` method within there, double click it and hit execute:

![image](https://user-images.githubusercontent.com/3372394/129694406-5f4692bc-9fe6-41c3-94e7-cd3e11f8d808.png)

**Comment by davidak** [comment]
Thanks. Nothing happens and nothing in the `dbus-monitor --session` output.

![Bildschirmfoto von 2021-08-17 11-29-13](https://user-images.githubusercontent.com/91113/129701119-7a1ad027-7e29-404f-9f34-9de3cb44af4b.png)

Is anything else than gala needed for it to work? Something that could be missing?

**Comment by davidmhewitt** [comment]
Well, I don't think your `dbus-monitor` is working properly, as that dbus call definitely travelled on the session bus, or else there would be an error in D-Feet. But nevermind, since we now know that it is Gala that is ignoring the request for some reason.

Take a look at the console logs/journal from when Gala started, there are a number of reasons why it would disable screen locking and it would have logged an error explaining why.

**Comment by davidak** [comment]
>when Gala started

at which line does that happen?

here is the full log since boot with debugging enabled https://gist.github.com/davidak/749b444fb38c753c4b02de3ba580f3c0

I see these journal lines on the live session, running in gnome boxes. Here locking also don't work.

```
Aug 17 09:46:45 elementary systemd[1490]: Started Evolution source registry.
Aug 17 09:46:45 elementary gala[2426]: clutter_actor_get_n_children: assertion 'CLUTTER_IS_ACTOR (self)' failed
Aug 17 09:46:45 elementary gala[2426]: gala_drag_drop_action_get_dragging: assertion 'self != NULL' failed
Aug 17 09:46:45 elementary dbus-daemon[1875]: [session uid=999 pid=1875] Activating via systemd: service name='org.gnome.evolution.dataserver.Calendar8' unit='>
Aug 17 09:46:45 elementary gala[2426]: STACK_OP_RAISE_ABOVE: sibling window 0x3000008 (io.element) not in stack
Aug 17 09:46:45 elementary systemd[1490]: Starting Evolution calendar service...
Aug 17 09:46:45 elementary gala[2426]: STACK_OP_RAISE_ABOVE: sibling window 0x3000008 not in stack
Aug 17 09:46:45 elementary bamfdaemon[2497]: Got an X error: BadWindow (invalid Window parameter)
Aug 17 09:46:45 elementary bamfdaemon[2497]: Got an X error: BadWindow (invalid Window parameter)
Aug 17 09:46:45 elementary bamfdaemon[2497]: Got an X error: BadWindow (invalid Window parameter)
Aug 17 09:46:45 elementary gala[2426]: STACK_OP_RAISE_ABOVE: sibling window 0x3000008 not in stack
```
i see the same lines on nixos

```
Aug 17 09:55:56 gaming .io.elementary.[2092]: DBusService.vala:64: Service registration suceeded
Aug 17 09:55:57 gaming .io.elementary.[2092]: RelevancyService.vala:166: Pushing launch event: application://atom.desktop [Atom]
Aug 17 09:55:59 gaming gala.desktop[2038]: Window manager warning: Ping serial 1248136 was reused for window 0x3a00001, previous use was for window 0x3a0000a.
Aug 17 09:55:59 gaming gala.desktop[2038]: Window manager warning: Ping serial 1248136 was reused for window 0x3a00001, previous use was for window 0x3a0000a.
Aug 17 09:55:59 gaming .gala-wrapped[2038]: clutter_actor_remove_child: assertion 'CLUTTER_IS_ACTOR (child)' failed
Aug 17 09:55:59 gaming .gala-wrapped[2038]: invalid (NULL) pointer instance
Aug 17 09:55:59 gaming .gala-wrapped[2038]: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
Aug 17 09:55:59 gaming .gala-wrapped[2038]: clutter_actor_remove_child: assertion 'CLUTTER_IS_ACTOR (child)' failed
Aug 17 09:55:59 gaming .gala-wrapped[2038]: invalid (NULL) pointer instance
Aug 17 09:55:59 gaming .gala-wrapped[2038]: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
Aug 17 09:55:59 gaming .gala-wrapped[2038]: clutter_actor_remove_child: assertion 'CLUTTER_IS_ACTOR (child)' failed
Aug 17 09:55:59 gaming .gala-wrapped[2038]: invalid (NULL) pointer instance
Aug 17 09:55:59 gaming .gala-wrapped[2038]: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
```

that looks like some issue, but not related to locking

**Comment by davidmhewitt** [comment]
You're right, I can't see anything in that log that would indicate a reason Gala would disable screen locking. Here are the relevant lines from the code:

https://github.com/elementary/gala/blob/87d51c50fbd5a28d2d0d6c3872b4e15e39c9131a/src/Widgets/ScreenShield.vala#L163-L190

It depends on:
- `logind` presenting a dbus interface at `org.freedesktop.login1`
- GNOME session manager presenting a presence bus at `org.gnome.SessionManager` path `/org/gnome/SessionManager/Presence`
- The `XDG_SEAT_PATH` variable being set
- The greeter presenting a dbus interface at `org.freedesktop.DisplayManager`

Do you have pantheon greeter configured and working in your setup? This has never been tested with other greeters.

**Comment by davidak** [comment]
- `org.freedesktop.login1` interface is there, provided by `systemd-logind`
- `org.gnome.SessionManager` interface is there, provided by `gnome-session-binary`

i see the methods "Logout" and "Reboot" there. should there be lock too?

![Bildschirmfoto von 2021-08-17 14-00-11](https://user-images.githubusercontent.com/91113/129722086-6c8b2c12-eb7c-4b7c-bdee-3f1f94e5ac38.png)

- `XDG_SEAT_PATH` is set, to `/org/freedesktop/DisplayManager/Seat0`
- `org.freedesktop.DisplayManager` is provided by `lightdm`

>Do you have pantheon greeter configured and working in your setup?

yes

https://github.com/NixOS/nixpkgs/blob/1a9eb85a922f23f4ef9b4eb5040b96c94131588e/nixos/modules/services/x11/display-managers/lightdm-greeters/pantheon.nix#L38

**Comment by davidak** [comment]
This is the last issue that is blocking us from updating pantheon.

How can we halp do get it fixed?

@davidmhewitt do you have an idea what causes this issue?

**Comment by davidmhewitt** [comment]
No idea, but provide me with a guide for setting up a test environment of NixOS with these packages in a VM and I'll have a look for you when I can.

**Comment by davidak** [comment]
What kind of VM does your system support? KVM/qemu or Virtualbox? I will try to provide you a disk image.

**Comment by davidmhewitt** [comment]
KVM/qemu would be fine.

Probably better that you just give me an image too since I've spent the last little while just getting increasingly frustrated trying to do it myself :sweat_smile:

**Comment by davidak** [comment]
I can't reproduce the issue in a new VM, only on my system that was updated from pantheon 5. The only difference should be configuration files in the home directory. I remember that the pantheon stylesheet was broken after i have used GNOME before on another system, so something like that could happen.

I try to install a NixOS VM with pantheon 5 and update it to pantheon 6 and see if i can reproduce the issue. It could take some time...

**Comment by bobby285271** [comment]
This is probably a very stupid question and I am not sure if someone have asked this before. By executing `gsettings get org.gnome.desktop.lockdown disable-lock-screen` can you get `false` on you system?

**Comment by davidak** [comment]
>I am not sure if I have asked this before

i think no

>can you get `false` on you system?

i get `true`

>then locking should work.

it does work then

**Comment by davidak** [comment]
@davidmhewitt do you have an idea why `org.gnome.desktop.lockdown disable-lock-screen` is set to `true` on the live image and on a system that was updated from elementary 5?

What can we do that this does not happen?

**Comment by davidmhewitt** [comment]
I've just had a look at why it's disabled on the live image and it's because it's overridden by a package called `casper` which gets included on the iso _but not the installed system_ to make the live session work. So once the system is installed, this will default to `false`.

The default value of that key is `false` and we've never overridden it or even offered a GUI way of changing it. The only reason Gala reads that key is to be respectful of it if someone has changed that preference in a GNOME session where that setting is available in the control center.

The only explanation I have for you is that you or something has changed that value from its default at some point and it's followed you through an upgrade. It never used to be used by the old light-locker based screen locking mechanism, so it wouldn't have had an effect previously.

**Comment by davidak** [comment]
It could be possible that i had GNOME installed on the system at some point.

**Comment by bobby285271** [comment]
> [a guide for setting up a test environment of NixOS with these packages in a VM](https://github.com/elementary/gala/issues/1214#issuecomment-914361262)

Though no longer needed by this issue, thanks to the recently released Nix 2.4 and the [flakes](https://nixos.wiki/wiki/Flakes) feature, setting such environment should be easier today. I just create a very simple [flake](https://github.com/bobby285271/test-pantheon) that will set up and launch a QEMU for you which provides latest NixOS (`nixos-unstable-small`) with default Pantheon experience NixOS provides (also some debug options is enabled for example `GNOME_SESSION_DEBUG` has been set).

So here are the instructions, I have tested the following steps on a Xubuntu 21.10 VM.

#### Step 1: Install Nix

Nix can be installed and used on any Linux distributions, not just NixOS. Nix 2.4 (latest stable) or later is needed for the flakes feature. To install Nix, simply follow the [installation guide](https://nixos.org/manual/nixos/stable/), that is, run the following command as a normal user (not `root`) with `sudo` permission:

```ShellSession
$ sh <(curl -L https://nixos.org/nix/install)
```

No questions should be asked during installation except asking password for `sudo`, and a `/nix` directory will be created automatically. After the installation finishes, log in again to make sure the necessary environment variables are set.

FYI, Nix can be uninstalled by running `rm -rf /nix` if you no longer want it.

#### Step 2: Build and launch the NixOS VM

This will do the rest of the things, just wait for it to download the needed packages, then QEMU will launch automatically, wait for LightDM to appear and login with username and password `test`. 

```ShellSession
$ nix --experimental-features "nix-command flakes" run github:bobby285271/test-pantheon
```

To test an ongoing Nixpkgs pull request, https://github.com/NixOS/nixpkgs/pull/145465 for example, run the following instead:

```ShellSession
$ nix --experimental-features "nix-command flakes" run github:bobby285271/test-pantheon --override-input nixpkgs github:NixOS/nixpkgs/pull/145465/head
```

To customize settings like [memorySize](https://github.com/bobby285271/test-pantheon/blob/a4987eb35b4458484a887f3d297c430eae08ab69/flake.nix#L32) for QEMU, it is possible to fork my [GitHub repo](https://github.com/bobby285271/test-pantheon) and do the changes, then replace `bobby285271/test-pantheon` with your GitHub repository.

![](https://user-images.githubusercontent.com/20080233/141946674-c36cd57f-32dc-444e-b3c1-a4f33ddd7a9a.png)
